### ISSUE TITLE
Transaction error when creating new tenant while celery backgroud tasks are being executed for other tenants

### ISSUE THREAD

**Issue by DanielFerreiraJorge** [tomturner/django-tenants#55]
Hi, I'm getting this when creating a new tenant if some celery tasks are being executed:
InternalError: current transaction is aborted, commands ignored until end of transaction block

I already have the latest version (master), which already have this PR merged:
https://github.com/CGenie/django-tenants/commit/e5b84089d8aa69be02c10ed2dc979f89de5c12f3

I tought this would solve the problem but apparently not. My system is very heavy and I have tenants that have background celery tasks being executed all the time, one task per tenant (each tenant has its own task)... this tasks connect to a webservice to download and save objects on the tenant schema (dozens of objects per second per tenant). One of these objects are very nested, composed by over 20 classes and, because of that, I save it in a transaction.atomic() block to ensure everything is created properly.

Now, this is IMPORTANT: If I modify my celery task and remove the transaction.atomic() when saving this object, my task still works (subject to integrity errors, of course) and I CAN CREATE THE TENANT without the error, as expected... so, apparently, there is some conflict when creating a tenant while atomically saving objects in other teants...

See also the sentry captured event that describes it in more detail:
![error](https://cloud.githubusercontent.com/assets/1966185/11425016/e4640616-9438-11e5-8bd7-4029370900b4.png)

**Comment by kosz85** [comment]
From my experience it's caused by not properly handled db error in your code. If you ran it as celery task it can be even broken transaction from previously ran task.
Check every `try` `except` block with DB action like creating, update, get_or_create, as well as any other error rised from DB that makes transaction broken. If you just pass it, and do not  do `rollback` and `savepoint_rollback` then error in such place can be a cause.
I piity you, as this is one of the worst  errors to find and fix, especially when it's at celery, as you need to analyze every piece of code, or close db connection on start ;)

**Comment by DanielFerreiraJorge** [comment]
Hi, thank you very much for you reply @kosz85 

I'm researching this problem for about a week now and it is driving me crazy... I think the problem is something else... I do not think there are broken transactions... the thing is that the celery tasks run LONG transactions (thousands of inserts per second) and when django_tenants try to create the tenants tables, there are some locks on foreign keys to tables on the public schema (which are being used by the celery tasks)... take a look at the excerpt below from the postgresql logs... the table base_municipio (which is in the public schema), for instance, is referenced in MANY tables and columns through the tenant schemas... there even are some tables that references it 6 times... so, when the celery task is transactionally inserting hundreds of rows in the tables that has foreign keys to the base_municipio table, dhango_tenants cannot create a schema with tables that reference base_municipio... please, somebody help... Thanks

2015-11-30 17:06:32 BRST [8913-1] postgres@daniel ERROR:  deadlock detected
2015-11-30 17:06:32 BRST [8913-2] postgres@daniel DETAIL:  Process 8913 waits for AccessExclusiveLock on relation 2004322 of database 16385; blocked by process 9083.
	Process 9083 waits for RowShareLock on relation 2004302 of database 16385; blocked by process 8913.
	Process 8913: ALTER TABLE "contas_empresa" ADD CONSTRAINT "contas_emp_municipio_id_34609ece9dc15eae_fk_base_municipio_ibge" FOREIGN KEY ("municipio_id") REFERENCES "base_municipio" ("ibge") DEFERRABLE INITIALLY DEFERRED
	Process 9083: COMMIT
2015-11-30 17:06:32 BRST [8913-3] postgres@daniel HINT:  See server log for query details.
2015-11-30 17:06:32 BRST [8913-4] postgres@daniel STATEMENT:  ALTER TABLE "contas_empresa" ADD CONSTRAINT "contas_emp_municipio_id_34609ece9dc15eae_fk_base_municipio_ibge" FOREIGN KEY ("municipio_id") REFERENCES "base_municipio" ("ibge") DEFERRABLE INITIALLY DEFERRED
2015-11-30 17:06:32 BRST [8913-5] postgres@daniel ERROR:  current transaction is aborted, commands ignored until end of transaction block
2015-11-30 17:06:32 BRST [8913-6] postgres@daniel STATEMENT:  SET search_path = arildo,public,extensions
2015-11-30 17:06:32 BRST [8913-7] postgres@daniel ERROR:  current transaction is aborted, commands ignored until end of transaction block
2015-11-30 17:06:32 BRST [8913-8] postgres@daniel STATEMENT:  SELECT EXISTS(SELECT 1 FROM pg_catalog.pg_namespace WHERE LOWER(nspname) = LOWER('arildo'))
2015-11-30 17:06:32 BRST [8913-9] postgres@daniel ERROR:  current transaction is aborted, commands ignored until end of transaction block
2015-11-30 17:06:32 BRST [8913-10] postgres@daniel STATEMENT:  SET search_path = arildo,public,extensions

**Comment by kosz85** [comment]
This error "current transaction is aborted, commands ignored until end of transaction block" is exactly what I said, and the deadlock is what cause it, as deadlock is db error. It should be handled, you have it at Sentry so i guess it's not handled and is breaking transaction.

As for deadlock itself. I create my tenants in celery, while other tenants are actively used, so there should be no problem with this. It's your db design causing it. You could use copy, or insert entries to temporary table and then copy it to proper place to not lock tables.
You can construct celery queue for tasks heavy on DB, run 1 at time, and schedule new tenants as importing, or... add creating new tenants (if requested) on the beginning of import task, and run it before importing. There are plenty of solutions.
But first I would see what exactly is waiting for what and what is locked and try to fix it. As probably you can do it, by changing order of transaction, spiting it into smaller pieces (for example 100-1000 rows), redesign tables and/or relations.

**Comment by kosz85** [comment]
I had similar issue few days ago. I don't know yet how to solve it - in my case I just run it on 1 process instead of using parallel approach, thought I think it was caused by migration of table structure and data migration in one migration. It's quite normal, that 1 migration is 1 transaction, having it more at once, doing same things and making them last longer (data migration) can be causing deadlocks. And for deadlock you just need 1 relation with shared model (User in my case).
I suggest to split migrations and any data migration in to separate migrations, this may solve some of problems.

**Comment by DanielFerreiraJorge** [comment]
This error happens whenever you have a foreign key to a table in the public schema (for instance, the tenant model)... 

If you use db_constraint=False in the django model, it does not happen because the FK will not be enforced at the database level... for me this is ok since no table in the public schema will have deletable records... well... it's a workaround

**Comment by kosz85** [comment]
But you can avoid it but using standard migration, without multiprocessing.

**Comment by tomturner** [comment]
A question are you using https://github.com/maciej-gol/tenant-schemas-celery or are you using celery direct.

**Comment by kosz85** [comment]
In my case I use multiprocessing for `migrate_schemas`, that were developed few weeks ago, I think we will need to improve it at some point of time, thought most of migrations pass just fine. Only that ones blocked by longer data migration have problems.
As for tasks, with witch I don't have any problem, we used our own solution, I was't aware of that one. We used new `BaseTask` class to store tenant at task apply time, and restore it at task.run time.

**Comment by DanielFerreiraJorge** [comment]
@kosz85 I do use standard migration, without multiprocessing... what is happening is that when the new tenant schema is being created and it contains a table with a foreign key to the public schema, the alter table tries to acquire an exclusive lock to the table in the public schema, and it can't, since the table is being used... this happens at the database level when creating the constraint. That is why I'm using db_constraint=False on these relations that point to a table in the public schema

@tomturner celery direct, but I truly do not think this is a celery problem, as I explained in the paragraph above... of course, I may be totally wrong...

**Comment by kosz85** [comment]
Thx, I understand now, I had different thought similar issue, with multiprocessing of migrations, but on migration level. Some kind of solution would be adding constraints in separate transaction, which would block less time, and retry it till success. It depends how heavy is your usage on shared table.

To bypass it without loosing constraints, you could think about duplicating pre created ready to use schema. Or pre creating set of such template schemas (for example 50) on downtime/migration time and just renaming it when it's needed. That way you shouldn't have exclusive locks at important times, as all constraints would be ready to use.
